File: string_matching_based_on_finite_automaton.py
import subprocess
import tempfile
import html
from collections import defaultdict
from collections import namedtuple

def pprinttable(rows):
    esc = lambda x: html.escape(str(x))
    sour = "<table border=1>"
    if len(rows) == 1:
        for i in range(len(rows[0]._fields)):
            sour += "<tr><th>%s<td>%s" % (esc(rows[0]._fields[i]), esc(rows[0][i]))
    else:
        sour += "<tr>" + "".join(["<th>%s" % esc(x) for x in rows[0]._fields])
        sour += "".join(["<tr>%s" % "".join(["<td>%s" % esc(y) for y in x]) for x in rows])
    with tempfile.NamedTemporaryFile(suffix=".html") as f:
        f.write(sour.encode("utf-8"))
        f.flush()
        print(
            subprocess
            .Popen(["w3m","-dump",f.name], stdout=subprocess.PIPE)
            .communicate()[0]
            .decode("utf-8")
            .strip()
        )

def dictstoobj(a):
  Row = namedtuple('Row',[x for x in sorted(list(a[0].keys()))])
  obj = [[y[1] for y in sorted(x[1].items())] for x in sorted(delta.items())]
  obj = [Row(*x) for x in obj]
  return obj

T="abcdabc"
alphabet = sorted(list(set(T)))
m = len(T)

f = lambda: defaultdict(f)
delta = defaultdict(f)

# http://wiki.algo.informatik.tu-darmstadt.de/index.php/String_matching_based_on_finite_automaton
# induction basis
for j in range(m):
	for c in alphabet:
		k = min(m-1, j+1)
		while k>0 and T[0:k] != T[j-k+1:j] + c: k = k - 1
		delta[j][c] = k

pprinttable(dictstoobj(delta))

S = "abcdabcdabcb"
n = len(S)
q = 0
R = []

# induction step
for i in range(1,n+1):
	q = delta[q][S[i-1]]
	if q == m-1: R.append(i-m+1)

print(R)

File: string_matching_based_on_finite_automaton.lua
function table.print(t, ...)
	for i=0,#t do
		local r = ""
		for _,j in ipairs(alphabet) do
			r = r .. ", " .. j .. "=" .. t[i][j]
		end
		print(i .. ": " .. r)
	end
end

T = "aabbccaa"
alphabet = {"a","b","c","d"}
m = #T

delta = {}

-- http://wiki.algo.informatik.tu-darmstadt.de/wiki.algo.informatik.tu-darmstadt.de/index.php/String_matching_based_on_finite_automaton.html
-- induction basis
delta[0] = {}
for _,c in ipairs(alphabet) do delta[0][c] = T:sub(1,1) == c and 1 or 0 end
for j=1,m do
        for _,c in ipairs(alphabet) do
                local k = math.min(m, j+1)
                while k > 0 and T:sub(1,k) ~= T:sub(j-k+2,j) .. c do
                   k = k - 1
                end
                if delta[j] == nil then delta[j] = {} end
                delta[j][c] = k
        end
end

table.print(delta)

S = "ccaabbccaa"
n = #S
q = 0
R = {}

-- induction step
for i=1,n do
        q = delta[q][S:sub(i,i)]
        if q == m then table.insert(R,i-m+1) end
end

print("size: " .. #R)
for _,p in ipairs(R) do
        print(p)
	local res = S:sub(p,p+#T-1)
	if T ~= res then print("error!" .. res) end
end
